plugin/http/httpstats/client.go

```diff
@@ -0,0 +1,158 @@
+	} {
+		fullname := qualify(c.name)
+		m, err := stats.NewMeasureInt64(fullname, c.desc, c.unit)
+		if err != nil {
+			log.Fatalf("failed to create measure %q: %s", fullname, err)
+		}
+		*c.measure = m
+		v, err := stats.NewView(fullname, c.desc, nil, m, &stats.CountAggregation{}, &stats.Cumulative{})
+		if err != nil {
+			log.Fatalf("failed to create view %q: %s", fullname, err)
+		}
+		*c.view = v
+	}
+
+	var err error
+	fullname := qualify("latency")
+	Latency, err = stats.NewMeasureFloat64(fullname, "End-to-end request latency", "microseconds")
+	if err != nil {
+		log.Fatalf("failed to create measure %q: %#v", fullname, err)
+	}
+	aggregation := &stats.DistributionAggregation{1, 100, 1000, 5000, 10000, 50000, 100000, 200000, 500000, 1000000}
+	LatencyDistribution, err = stats.NewView(fullname, "End-to-end request latency", nil, Latency, aggregation, &stats.Cumulative{})
+
+	for _, c := range []struct {
+		key  *tag.Key
+		name string
+	}{
+		{key: &Host, name: "host"},
+		{key: &StatusCode, name: "status_code"},
+		{key: &Path, name: "path"},
+		{key: &Method, name: "method"},
+	} {
+		fullname := qualify(c.name)
+		m, err := tag.NewKey(fullname)
+		if err != nil {
+			log.Fatalf("failed to validate tag key %q: %s", fullname, err)
+		}
+		*c.key = m
+	}
+}
+
+func qualify(suffix string) string {
+	return fmt.Sprintf("opencensus.io/http/client/%s", suffix)
+}
+
+// Transport is an http.RoundTripper that collects stats for the outgoing requests.
+type Transport struct {
+	// Base represents the underlying roundtripper that does the actual requests.
+	// If none is given, http.DefaultTransport is used.
+	//
+	// If base HTTP roundtripper implements CancelRequest,
+	// the returned round tripper will be cancelable.
+	Base http.RoundTripper
+}
+
+func (t Transport) base() http.RoundTripper {
+	if t.Base != nil {
+		return t.Base
+	}
+	return http.DefaultTransport
+}
+
+func (t Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	tags, _ := tag.NewMap(req.Context(),
+		tag.Upsert(Host, req.URL.Host),
+		tag.Upsert(Path, req.URL.Path),
+		tag.Upsert(Method, req.Method))
+	ctx := tag.NewContext(req.Context(), tags)
+	ctx = httptrace.WithClientTrace(ctx, trace(ctx))
+	req = req.WithContext(ctx)
+	var (
+		resp *http.Response
+		err  error
+	)
+	start := time.Now()
+	defer func() {
+		finishAndRecordStats(ctx, start, resp, err)
+	}()
+	stats.Record(ctx, Started.M(1))
+	resp, err = t.base().RoundTrip(req)
+	return resp, err
+}
+
+func finishAndRecordStats(ctx context.Context, start time.Time, resp *http.Response, err error) {
+	if resp != nil {
+		tags, _ := tag.NewMap(ctx, tag.Upsert(StatusCode, strconv.Itoa(resp.StatusCode)))
+		ctx = tag.NewContext(ctx, tags)
+	}
+	stats.Record(ctx, Latency.M(float64(time.Since(start))/float64(time.Microsecond)))
+}
+
+func trace(ctx context.Context) *httptrace.ClientTrace {
+	prev := httptrace.ContextClientTrace(ctx)
```

Comment:
Shouldn't we just install a new ClientTrace implementation for each request instead?

Reply:
the user might have installed their own ClientTrace, and I wanted that to still work